File: lib/lambda/rules/aws_cis/aws_cis_3_4.ts
import { cloudTrailRecord, uguisuRule, detection } from "../../models";

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.4-remediation

export class rule extends uguisuRule {
  constructor() {
    super({
      id: "aws_cis_3.4",
      title: "IAM policy changes",
      description:
        "AWS CIS benchmark 3.4 recommend to ensure a log metric filter and alarm exist for IAM policy changes",
      severity: "medium",
    });
  }

  detect(record: cloudTrailRecord): detection | null {
    const targetEvents: { [key: string]: boolean } = {
      DeleteGroupPolicy: true,
      DeleteRolePolicy: true,
      DeleteUserPolicy: true,
      PutGroupPolicy: true,
      PutRolePolicy: true,
      PutUserPolicy: true,
      CreatePolicy: true,
      DeletePolicy: true,
      CreatePolicyVersion: true,
      DeletePolicyVersion: true,
      AttachRolePolicy: true,
      DetachRolePolicy: true,
      AttachUserPolicy: true,
      DetachUserPolicy: true,
      AttachGroupPolicy: true,
      DetachGroupPolicy: true,
    };

    if (targetEvents[record.eventName] === true) {
      return {
        rule: this,
        event: record,
      };
    }

    return null;
  }
}

File: pkg/rules/aws_cis_3_4.go
package rules

import (
	"github.com/m-mizutani/uguisu/pkg/models"
)

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.4-remediation
type awsCIS3_4 struct {
	targetEvents map[string]bool
}

func newAwsCIS3_4() models.Rule {
	return &awsCIS3_4{
		targetEvents: map[string]bool{
			"DeleteGroupPolicy":   true,
			"DeleteRolePolicy":    true,
			"DeleteUserPolicy":    true,
			"PutGroupPolicy":      true,
			"PutRolePolicy":       true,
			"PutUserPolicy":       true,
			"CreatePolicy":        true,
			"DeletePolicy":        true,
			"CreatePolicyVersion": true,
			"DeletePolicyVersion": true,
			"AttachRolePolicy":    true,
			"DetachRolePolicy":    true,
			"AttachUserPolicy":    true,
			"DetachUserPolicy":    true,
			"AttachGroupPolicy":   true,
			"DetachGroupPolicy":   true,
		},
	}
}

func (x *awsCIS3_4) ID() string                { return "aws_cis_3.4" }
func (x *awsCIS3_4) Title() string             { return "IAM policy changes" }
func (x *awsCIS3_4) Severity() models.Severity { return models.SeverityMedium }
func (x *awsCIS3_4) Description() string {
	return "AWS CIS benchmark 3.4 recommend to ensure a log metric filter and alarm exist for IAM policy changes"
}
func (x *awsCIS3_4) Match(record *models.CloudTrailRecord) bool {
	return x.targetEvents[record.EventName] == true
}
